### ISSUE TITLE
Port namedAny and friends and filenameToModuleName (from twisted.python.reflect) to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5914]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5914|
|Type|enhancement|
|Created|2012-08-24 18:04:53Z|
|Branch|https://github.com/twisted/twisted/tree/namedAny-python3-5914-2|

`namedAny` and `filenameToModuleName` are used by trial, and so it would be good to port them to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5914 5914
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_namedAny_python3_5914_2 branches/namedAny-python3-5914-2
branch_author__itamar itamar
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1345831493000000 1345831493000000
changetime__1346250541000000 1346250541000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35382]) Branching to 'namedAny-python3-5914'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [35413]) Branching to 'namedAny-python3-5914-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, ready for review.

  1. Tests are unmodified, except they refer to different modules/objects.
  1. Code is unmodified, except for switch to `reraise()`.
  1. Added `reraise()` to twisted.python.compat, with tests.
  1. I ported a few extra `named*` just because they were in same test suite and it's just a matter of copy/paste; if there's issues, namedAny is the only important one and we can drop them with a little more work.
  1. `filenameToModule` supports Unicode filenames on Python 3. The implementation may break non-ASCII module names, but that was supposed to be a Python 3.3 feature and I don't see it in changelog so maybe that never happened. And in any case no one will have written those yet.
  1. Tests for `ihooks` started failing after changes in item 1 above. I get the impression ihooks changed in 2.6, our support was broken, but our tests never triggered it. Certainly the implementation didn't change. So, I just skipped the tests (they can be todo instead if you care).
  1. Buildbot run started here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/namedAny-python3-5914-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

  1. There are some new pyflakes warnings that should be fixed
  1. There's no test coverage for the `__traceback__` check in `reraise`.  More generally, I'm not sure `reraise` needs to handle the case where a traceback isn't passed in?  The only use of it is the two argument form with a non-`None` traceback.
  1. I think `LookupsTestCase` needs to be deleted from the global namespace in `test_compat.py` after it is used to create the ihooks subclass.

Please merge when those three points are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst set status to `closed`|
|-|-|

(In [35441]) Merge namedAny-python3-5914-2

Author: itamar
Review: exarkun
Fixes: [#5914](https://github.com/twisted/twisted/issues/5914)

More functionality from twisted.python.reflect works on Python 3.
Also, added a reraise() functions for re-raising exceptions across both Python 2 and Python 3.
